mountpoint-s3-client/src/object_client.rs

```diff
@@ -428,6 +428,11 @@ pub trait GetObjectRequest:
 {
     type ClientError: std::error::Error + Send + Sync + 'static;
 
+    /// Get the object's user defined metadata.
+    /// If the metadata has already been read, return immediately. Otherwise, resolve the future
+    /// when they're read.
+    async fn get_object_metadata(&mut self) -> Result<HashMap<String, String>, Self::ClientError>;
```

Comment:
Let's make this callable after the whole request was read, which is when the headers are certainly available and fits our use-case. This would mean the following signature:
```rust
async fn get_object_metadata(self: Pin<&mut Self>) -> Result<HashMap<String, String>, Self::ClientError>;
```

```rust
// failure_client.rs -- body changes
async fn get_object_metadata(self: Pin<&mut Self>) -> Result<HashMap<String, String>, Self::ClientError> {
    let this = self.project();
    this.request.get_object_metadata().await
}
```

```rust 
// s3_crt_client.rs -- no changes in body
async fn get_object_metadata(self: Pin<&mut Self>) -> Result<HashMap<String, String>, Self::ClientError> {
    self.object_metadata_cb
        .get_unpin()
        .await
        .clone()
        .map_err(|_| S3RequestError::RequestCanceled)
}
```

Also let's add a test for this scenario:
```rust 
let mut result = client
    .get_object(&bucket, &key, None, None)
    .await
    .expect("get_object should succeed");

let expected = &body;
pin_mut!(result);
while let Some(_) = result.next().await {
}
let actual_metadata = result.get_object_metadata().await.expect("Should return metadata");
assert_eq!(actual_metadata, metadata);
```

Reply:
I decided to migrate to the pin approach anyway, but I don't want to make it require reading from the stream.